File: java/src/main/java/vsdk/toolkit/io/image/DkColor.java
package vsdk.toolkit.io.image;

/**
Routines using pixel color values / color calculations.
     12/31/85

Two color representations are used, one for calculation and
another for storage.  Calculation is done with three floats
for speed.  Stored color values use 4 bytes which contain
three single byte mantissas and a common exponent.
*/

import java.io.OutputStream;
import vsdk.toolkit.common.ColorRgb;
import vsdk.toolkit.common.CppReAlloc;
import vsdk.toolkit.io.PersistenceElement;

public class DkColor {
    private static final int RED = 0;
    private static final int GREEN = 1;
    private static final int BLUE = 2;
    private static final int EXP = 3;
    private static final int COL_XS = 128;
    private static final int MINIMUM_SCAN_LINE_LENGTH = 8;
    private static final int MAXIMUM_SCAN_LINE_LENGTH = 0x7fff;
    private static final int MINIMUM_RUN_LENGTH = 4;

    private static byte[] temporaryBuffer = null;
    private static int temporaryBufferLength = 0;

    private static void writeByte(OutputStream stream, int value) {
        if (stream == null) {
            return;
        }
        try {
            stream.write(value & 0xFF);
        }
        catch (Exception ignored) {
        }
    }

    /**
Get a temporary buffer
*/
    private static byte[] tempBuffer(int length) {
        if (length > temporaryBufferLength) {
            if (temporaryBufferLength > 0) {
                temporaryBuffer = CppReAlloc.reAlloc(
                    temporaryBuffer,
                    temporaryBufferLength,
                    length);
            }
            else {
                temporaryBuffer = new byte[length];
            }
            temporaryBufferLength = temporaryBuffer == null ? 0 : length;
        }
        return temporaryBuffer;
    }

    /**
Write out a byte color scanline
*/
    private static int writeByteColors(byte[] scanline, int len, OutputStream outputStream) {
        int cnt = 0;
        int c2;

        if (outputStream == null) {
            return -1;
        }

        if (len < MINIMUM_SCAN_LINE_LENGTH || len > MAXIMUM_SCAN_LINE_LENGTH) {
            // OOBs, write out flat
            PersistenceElement.writeBytes(outputStream, scanline, len * 4);
            return 0;
        }

        // Put magic header
        writeByte(outputStream, 2);
        writeByte(outputStream, 2);
        writeByte(outputStream, len >> 8);
        writeByte(outputStream, len & 255);

        // Put components separately
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < len; j += cnt) {
                // Find next run
                int beg;

                for (beg = j; beg < len; beg += cnt) {
                    for (cnt = 1;
                         cnt < 127
                         && beg + cnt < len
                         && scanline[(beg + cnt) * 4 + i] == scanline[beg * 4 + i];
                         cnt++) {
                    }
                    if (cnt >= MINIMUM_RUN_LENGTH) {
                        // Long enough
                        break;
                    }
                }

                if (beg - j > 1 && beg - j < MINIMUM_RUN_LENGTH) {
                    c2 = j + 1;
                    while (c2 < beg && scanline[c2 * 4 + i] == scanline[j * 4 + i]) {
                        c2++;
                        if (c2 == beg) {
                            // Short run
                            writeByte(outputStream, 128 + beg - j);
                            writeByte(outputStream, scanline[j * 4 + i]);
                            j = beg;
                            break;
                        }
                    }
                }
                while (j < beg) {
                    // Write out non-run
                    if ( (c2 = beg - j) > 128 ) {
                        c2 = 128;
                    }
                    writeByte(outputStream, c2);
                    while (c2-- > 0) {
                        writeByte(outputStream, scanline[j * 4 + i]);
                        j++;
                    }
                }
                if (cnt >= MINIMUM_RUN_LENGTH) {
                    // Write out run
                    writeByte(outputStream, 128 + cnt);
                    writeByte(outputStream, scanline[beg * 4 + i]);
                }
                else {
                    cnt = 0;
                }
            }
        }
        return 0;
    }

    /**
Assign a short color value
*/
    private static void setByteColors(byte[] scanline, int base, double r, double g, double b)
    {
        double d = r > g ? r : g;
        if (b > d) {
            d = b;
        }

        if (d < 0) {
            scanline[base + RED] = 0;
            scanline[base + GREEN] = 0;
            scanline[base + BLUE] = 0;
            scanline[base + EXP] = 0;
            return;
        }

        final int e = Math.getExponent(d) + 1;
        final double normalized = Math.scalb(d, -e);
        d = normalized * 255.9999 / d;

        scanline[base + RED] = (byte)(int)(r * d);
        scanline[base + GREEN] = (byte)(int)(g * d);
        scanline[base + BLUE] = (byte)(int)(b * d);
        scanline[base + EXP] = (byte)(int)(e + COL_XS);
    }

    /**
Write out a scanline
*/
    public static int writeScan(float[][] scanline, int len, OutputStream outputStream)
    {
        // Get scanline buffer
        byte[] colorScan = tempBuffer(len * 4);
        if (colorScan == null) {
            return -1;
        }

        // Convert scanline
        for (int n = 0; n < len; n++) {
            int base = n * 4;
            setByteColors(colorScan, base, scanline[n][RED], scanline[n][GREEN], scanline[n][BLUE]);
        }
        return writeByteColors(colorScan, len, outputStream);
    }

    public static int writeScan(ColorRgb[] scanline, int len, OutputStream outputStream)
    {
        // Get scanline buffer
        byte[] colorScan = tempBuffer(len * 4);
        if (colorScan == null) {
            return -1;
        }

        // Convert scanline
        for (int n = 0; n < len; n++) {
            int base = n * 4;
            setByteColors(colorScan, base, scanline[n].r, scanline[n].g, scanline[n].b);
        }
        return writeByteColors(colorScan, len, outputStream);
    }

    public static void freeBuffer() {
        if (temporaryBuffer != null) {
            temporaryBuffer = null;
            temporaryBufferLength = 0;
        }
    }
}

File: turboc/src/io/image/Dkcolor.cpp
#include "java/lang/Math.h"
#include "common/CppReAlloc.h"
#include "io/wrapper/PersistenceElement.h"
#include "io/image/Dkcolor.h"

BYTE *DkColor::temporaryBuffer = NULL;
unsigned int DkColor::temporaryBufferLength = 0;

void
DkColor::writeByte(OutputStream *stream, int value) {
    if ( stream == NULL ) {
        return;
    }
    stream->write(value & 0xFF);
}

/**
Get a temporary buffer
*/
BYTE *
DkColor::tempBuffer(unsigned int length) {
    if ( length > temporaryBufferLength ) {
        if ( temporaryBufferLength > 0 ) {
            temporaryBuffer = CppReAlloc::reAlloc(
                temporaryBuffer,
                ((int)(temporaryBufferLength)),
                ((int)(length)));
        } else {
            temporaryBuffer = new BYTE[length];
        }
        temporaryBufferLength = temporaryBuffer == NULL ? 0 : length;
    }
    return temporaryBuffer;
}

/**
Write out a byte color scanline
*/
int
DkColor::writeByteColors(BYTE_COLOR *scanline, int len, OutputStream *outputStream) {
    int cnt = 0;
    int c2;

    if ( outputStream == NULL ) {
        return -1;
    }

    if ( len < MINIMUM_SCAN_LINE_LENGTH || len > MAXIMUM_SCAN_LINE_LENGTH ) {
        // OOBs, write out flat
        PersistenceElement::writeBytes(
            *outputStream,
            ((unsigned char *)(scanline)),
            ((int)(sizeof(BYTE_COLOR) * len)));
        return 0;
    }

    // Put magic header
    writeByte(outputStream, 2);
    writeByte(outputStream, 2);
    writeByte(outputStream, len >> 8);
    writeByte(outputStream, len & 255);

    // Put components separately
    for ( int i = 0; i < 4; i++ ) {
        for ( int j = 0; j < len; j += cnt ) {
            // Find next run
            int beg;

            for ( beg = j; beg < len; beg += cnt ) {
                for ( cnt = 1; cnt < 127 && beg + cnt < len && scanline[beg + cnt][i] == scanline[beg][i]; cnt++ ) {}
                if ( cnt >= MINIMUM_RUN_LENGTH ) {
                    // Long enough
                    break;
                }
            }

            if ( beg - j > 1 && beg - j < MINIMUM_RUN_LENGTH ) {
                c2 = j + 1;
                while ( scanline[c2++][i] == scanline[j][i] ) {
                    if ( c2 == beg ) {
                        // Short run
                        writeByte(outputStream, 128 + beg - j);
                        writeByte(outputStream, scanline[j][i]);
                        j = beg;
                        break;
                    }
                }
            }
            while ( j < beg ) {
                // Write out non-run
                if ( (c2 = beg - j) > 128 ) {
                    c2 = 128;
                }
                writeByte(outputStream, c2);
                while ( c2-- ) {
                    writeByte(outputStream, scanline[j++][i]);
                }
            }
            if ( cnt >= MINIMUM_RUN_LENGTH ) {
                // Write out run
                writeByte(outputStream, 128 + cnt);
                writeByte(outputStream, scanline[beg][i]);
            } else {
                cnt = 0;
            }
        }
    }
    return 0;
}

/**
Assign a short color value
*/
void
DkColor::setByteColors(BYTE_COLOR color, double r, double g, double b)
{
    double d = r > g ? r : g;
    if ( b > d ) {
        d = b;
    }

    if ( d < 0 ) {
        color[RED] = 0;
        color[GREEN] = 0;
        color[BLUE] = 0;
        color[EXP] = 0;
        return;
    }

    const int e = Math::getExponent(d) + 1;
    const double normalized = Math::scalb(d, -e);
    d = normalized * 255.9999 / d;

    color[RED] = ((unsigned char)(r * d));
    color[GREEN] = ((unsigned char)(g * d));
    color[BLUE] = ((unsigned char)(b * d));
    color[EXP] = ((unsigned char)(e + COL_XS));
}

/**
Write out a scanline
*/
int
DkColor::writeScan(DK_COLOR *scanline, int len, OutputStream *outputStream)
{
    // Get scanline buffer
    BYTE *byteArray = tempBuffer(len * sizeof(BYTE_COLOR));
    BYTE_COLOR *sp = ((BYTE_COLOR *)(byteArray));
    if ( sp == NULL ) {
        return -1;
    }
    BYTE_COLOR *colorScan = sp;

    // Convert scanline
    for ( int n = 0; n < len; n++ ) {
        setByteColors(sp[n], scanline[n][RED], scanline[n][GREEN], scanline[n][BLUE]);
    }
    return writeByteColors(colorScan, len, outputStream);
}

void
DkColor::freeBuffer() {
    if ( temporaryBuffer != NULL ) {
        delete[] temporaryBuffer;
        temporaryBuffer = NULL;
        temporaryBufferLength = 0;
    }
}
